enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md

```diff
@@ -0,0 +1,363 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedClusterScalarList is a collection of managed cluster scalar.
+type ManagedClusterScalarList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is a list of managed clusters
+	Items []ManagedClusterScalar `json:"items"`
+}
+```
+
+### Let placement prioritizer maintain the lifecycle of the CR.
+- Create
+
+The creation of ManagedClusterScalar CR is handled by each prioritizer plugin on demand. As every update in ManagedClusterScalar CR will introduce an api call, that will bring potential performance presure to api server. So we only want to create ManagedClusterScalar CRs for enabled prioritizer in filterd cluster namespace.
+
+The placement schedule framework will filter avaliable clusters as first step.
+Then the framework will call each prioritizer one by one. If the prioritizer is enable in placement yaml, it will create a corresponding ManagedClusterScalar CR in filtered cluster namespace as below. 
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: ManagedClusterScalar
+metadata:
+  name: {PrioritizerName}
+  namespace: cluster1
+spec:
+  prioritizerName: {PrioritizerNameInPlacement}
+```
+The name is the prioritizer name.
+The namespace is the filtered cluster namespace.
+The spec.prioritizerName is the prioritizer name used in placement yaml. 
+
+The code change includes add PreScore() in Prioritizer interface, in which to create ManagedClusterScalar CR before Score().
+```go
+// Prioritizer defines a prioritizer plugin that score each cluster. The score is normalized
+// as a floating betwween 0 and 1.
+type Prioritizer interface {
+	Plugin
+
+	// PreScore() do some prepare work before Score().
+	// For example, creating ManagedClusterScalar CR before Score().
+	PreScore(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) error
+
+	// Score gives the score to a list of the clusters, it returns a map with the key as
+	// the cluster name.
+	Score(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error)
+}
+```
+
+- Delete
+
+The deletion of ManagedClusterScalar CR is handled by the framework (out of all the prioritizer plugins). The ManagedClusterScalar CR is created by each prioritizer plugin on demand, and the CR should be shared by some placements cross namespaces. When the ManagedClusterScalar CR is not used by any placement, need to clean it to avoid unnecessary api call.
+
+There will be a GC() funcion in the framework, being called every 10 minutes. Go through all the Placment CRs and ManagedClusterScalar CRs, and clean not used ManagedClusterScalar CRs.
+
```

Comment:
we should also notify agent and controller about this, so that they will not report more data.

Reply:
Is there a timeout for the CR? E.g. for the cpu usage, if we cannot get new data anymore, how can we handle this case?

Reply:
> we should also notify agent and controller about this, so that they will not report more data.

When unused ManagedClusterScalar CR is removed by GC(), the agent won't watch and update scalar to status anymore.

> Is there a timeout for the CR? E.g. for the cpu usage, if we cannot get new data anymore, how can we handle this case?

Yes, the API has `ValidUntil` designed in CR status. 
```
//ManagedClusterScalarStatus represents the current status of ManagedClusterScalar.
type ManagedClusterScalarStatus struct {
	// Conditions contain the different condition statuses for this managed cluster scalar.
	// +optional
	Conditions []metav1.Condition `json:"conditions"`
	// Scalar contains a scalable value of this managed cluster.
	// +optional
	Scalar int64 `json:"scalar,omitempty"`
	// ValidUntil defines the time this scalar is valid.
	// After this time, the scalar is considered to be invalid by placement. nil means never expire.
	// The controller ownning this resource should keep the scalar up-to-date.
	// +optional
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Type=string
	// +kubebuilder:validation:Format=date-time
	ValidUntil *metav1.Time `json:"validUntil" protobuf:"bytes,4,opt,name=lastTransitionTime"`
}
```